### PR TITLE
ocamlPackages: irmin 2.2.0 → 2.4.0; git: 2.13 → 3.2.0; and related changes

### PR DESCRIPTION
**Pull request by sternenseemann**
###### Motivation for this change

This pull request updates `git` and `irmin` to their latest versions and adds some missing packages for them (most notably their `-mirage*` variants, reference #23955).

All other changes are related to this as these updates and new packaages are (indirectly)  required by these updates. Some of them can be reviewed and merged separately and their commits are also PR'd separately:

* ~~#109758~~
* ~~#109757~~
* ~~#109756~~
* ~~#109621~~
* ~~#109439~~

#### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by sternenseemann** [comment]
Result of `nixpkgs-review pr 110184` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>46 packages built:</summary>
  <ul>
    <li>ocamlPackages.bheap</li>
    <li>ocamlPackages.carton</li>
    <li>ocamlPackages.carton-git</li>
    <li>ocamlPackages.carton-lwt</li>
    <li>ocamlPackages.cohttp-mirage</li>
    <li>ocamlPackages.conduit-mirage</li>
    <li>ocamlPackages.decompress</li>
    <li>ocamlPackages.duff</li>
    <li>ocamlPackages.either</li>
    <li>ocamlPackages.emile</li>
    <li>ocamlPackages.encore</li>
    <li>ocamlPackages.git</li>
    <li>ocamlPackages.git-cohttp</li>
    <li>ocamlPackages.git-cohttp-mirage</li>
    <li>ocamlPackages.git-cohttp-unix</li>
    <li>ocamlPackages.git-unix</li>
    <li>ocamlPackages.hxd</li>
    <li>ocamlPackages.imagelib</li>
    <li>ocamlPackages.index</li>
    <li>ocamlPackages.irmin</li>
    <li>ocamlPackages.irmin-chunk</li>
    <li>ocamlPackages.irmin-containers</li>
    <li>ocamlPackages.irmin-fs</li>
    <li>ocamlPackages.irmin-git</li>
    <li>ocamlPackages.irmin-graphql</li>
    <li>ocamlPackages.irmin-http</li>
    <li>ocamlPackages.irmin-layers</li>
    <li>ocamlPackages.irmin-mem</li>
    <li>ocamlPackages.irmin-mirage</li>
    <li>ocamlPackages.irmin-mirage-git</li>
    <li>ocamlPackages.irmin-mirage-graphql</li>
    <li>ocamlPackages.irmin-pack</li>
    <li>ocamlPackages.irmin-test</li>
    <li>ocamlPackages.irmin-unix</li>
    <li>ocamlPackages.pecu</li>
    <li>ocamlPackages.ppx_irmin</li>
    <li>ocamlPackages.ppx_repr</li>
    <li>ocamlPackages.progress</li>
    <li>ocamlPackages.repr</li>
    <li>ocamlPackages.semaphore-compat</li>
    <li>ocamlPackages.terminal_size</li>
    <li>ocamlPackages.tls-mirage</li>
    <li>ocamlPackages.vchan</li>
    <li>ocamlPackages.xenstore</li>
    <li>ocamlPackages.xenstore-tool</li>
    <li>ocamlPackages.xenstore_transport</li>
  </ul>
</details>

**Comment by sternenseemann** [comment]
Getting the following build error for `wodan-irmin`:

```
      ocamlc src/wodan-irmin/.wodan_irmin.objs/byte/wodan_irmin.{cmo,cmt} (exit 2)
(cd _build/default && /nix/store/11xyy5msbabcli88gfrp5dh64n1mp4hb-ocaml-4.10.0/bin/ocamlc.opt -w -40 -g -g -bin-annot -I src/wodan-irmin/.wodan_irmin.objs/byte -I /nix/store/1myqnkifql7mc8ic4sd98ds5q69vbc3f-ocaml4.10.0-wodan-unstable-2020-11-20/lib/ocaml/4.10.0/site-lib/wodan -I /nix/store/1s0nb4rfzhwn0i0xb9hfva3hqnp28ar0-ocaml4.10.0-irmin-chunk-2.3.0/lib/ocaml/4.10.0/site-lib/irmin-chunk -I /nix/store/2pz330gg5xbq21lxy3358a032mlc4fxl-ocaml4.10.0-logs-0.7.0/lib/ocaml/4.10.0/site-lib/logs -I /nix/store/5hnzrqncm02l5r73vfdpd0q3ww1qlqpm-ocaml4.10.0-irmin-2.3.0/lib/ocaml/4.10.0/site-lib/irmin -I /nix/store/865njgmzkbbimsbmss6rqlp35lhp1bfn-ocaml4.10.0-irmin-git-2.3.0/lib/ocaml/4.10.0/site-lib/irmin-git -I /nix/store/kbgylgqi3f0vkwlcil7jm7hsnmyc14jm-ocaml4.10.0-fmt-0.8.9/lib/ocaml/4.10.0/site-lib/fmt -I /nix/store/nz5g64nldcgx87ivr4ddllrs3shwpjls-ocaml4.10.0-lwt-5.3.0/lib/ocaml/4.10.0/site-lib/lwt -I /nix/store/z2yif5iwmmanjnravz2b5x5mkn3ywjx0-ocaml4.10.0-mirage-block-2.0.1/lib/ocaml/4.10.0/site-lib/mirage-block -intf-suffix .ml -no-alias-deps -o src/wodan-irmin/.wodan_irmin.objs/byte/wodan_irmin.cmo -c -impl src/wodan-irmin/wodan_irmin.pp.ml)
File "src/wodan-irmin/wodan_irmin.ml", line 285, characters 7-19:
285 |   (V : Irmin.Type.S)
             ^^^^^^^^^^^^
Error: The module Irmin.Type is an alias for module Repr, which is missing
    ocamlopt src/wodan-irmin/.wodan_irmin.objs/native/wodan_irmin.{cmx,o} (exit 2)
(cd _build/default && /nix/store/11xyy5msbabcli88gfrp5dh64n1mp4hb-ocaml-4.10.0/bin/ocamlopt.opt -w -40 -g -g -g -O3 -I src/wodan-irmin/.wodan_irmin.objs/byte -I src/wodan-irmin/.wodan_irmin.objs/native -I /nix/store/1myqnkifql7mc8ic4sd98ds5q69vbc3f-ocaml4.10.0-wodan-unstable-2020-11-20/lib/ocaml/4.10.0/site-lib/wodan -I /nix/store/1s0nb4rfzhwn0i0xb9hfva3hqnp28ar0-ocaml4.10.0-irmin-chunk-2.3.0/lib/ocaml/4.10.0/site-lib/irmin-chunk -I /nix/store/2pz330gg5xbq21lxy3358a032mlc4fxl-ocaml4.10.0-logs-0.7.0/lib/ocaml/4.10.0/site-lib/logs -I /nix/store/5hnzrqncm02l5r73vfdpd0q3ww1qlqpm-ocaml4.10.0-irmin-2.3.0/lib/ocaml/4.10.0/site-lib/irmin -I /nix/store/865njgmzkbbimsbmss6rqlp35lhp1bfn-ocaml4.10.0-irmin-git-2.3.0/lib/ocaml/4.10.0/site-lib/irmin-git -I /nix/store/kbgylgqi3f0vkwlcil7jm7hsnmyc14jm-ocaml4.10.0-fmt-0.8.9/lib/ocaml/4.10.0/site-lib/fmt -I /nix/store/nz5g64nldcgx87ivr4ddllrs3shwpjls-ocaml4.10.0-lwt-5.3.0/lib/ocaml/4.10.0/site-lib/lwt -I /nix/store/z2yif5iwmmanjnravz2b5x5mkn3ywjx0-ocaml4.10.0-mirage-block-2.0.1/lib/ocaml/4.10.0/site-lib/mirage-block -intf-suffix .ml -no-alias-deps -o src/wodan-irmin/.wodan_irmin.objs/native/wodan_irmin.cmx -c -impl src/wodan-irmin/wodan_irmin.pp.ml)
File "src/wodan-irmin/wodan_irmin.ml", line 285, characters 7-19:
285 |   (V : Irmin.Type.S)
             ^^^^^^^^^^^^
Error: The module Irmin.Type is an alias for module Repr, which is missing
```

`repr` should be propagated, so I'm not quite sure what the issue here is. @ehmry any ideas?

**Comment by vbgl** [comment]
Merged the addition of `terminal_size` as b4668b2699a72fc3ddb4ea8c0ef67e993c9bef1b.

**Comment by ehmry** [comment]
@sternenseemann, don't know, unfortunately I don't know much of OCaml.

**Comment by sternenseemann** [comment]
Patching in a reference to repr in the dune file fixes this issue, but there are further incompatibilities with irmin 2.3.0 breaking the build.

I'm uncertain what to do about this since we can't update `wodan` to current master as it depends on a fork of `mirage-block` which creates additional packaging problems. So even if this gets fixed on master it doesn't really help us.

I'm half-tempted to mark `wodan-irmin` as broken and wait this out?

**Comment by ehmry** [comment]
So be it, wodan-irmin isn't a dependency of anything yet.

**Comment by vbgl** [comment]
Merged the addition of `progress` as fdc9d99179da74d77ea060f036f4b2539b65efba.

**Comment by sternenseemann** [comment]
I've added the updates to irmin 2.4.0, carton 0.2.0 and git 2.3.0 in here as well. As separate commits because this may help reviewing, but I can also squash the updates together if you want.

File: pkgs/development/ocaml-modules/carton/default.nix [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    checkseum
+    logs
+    psq
+    fmt
+    digestif
+    mmap
+    result
+    rresult
+    fpath
+    bos
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/carton/default.nix:46]: These libraries need not be propagated.

**Review by vbgl — CHANGES_REQUESTED** [review]
`decompress` won’t build with OCaml < 4.07:

>  File "lib/lzo.ml", line 113, characters 4-34:
  Error: Unknown builtin primitive "%caml_bytes_set32u"

File: pkgs/development/ocaml-modules/carton/default.nix (hunk 1 above)
Comment (sternenseemann) [reply pkgs/development/ocaml-modules/carton/default.nix]: Right, fixed.

**Comment by sternenseemann** [comment]
> `decompress` won’t build with OCaml < 4.07:

It doesn't support those versions anymore, I've now adjusted `minimumOCamlVersion` accordingly. It doesn't seem to me that we need `decompress` for 4.06 and below.

File: pkgs/development/ocaml-modules/irmin/unix.nix [hunk 2]
```diff
@@ -12,8 +12,10 @@ buildDunePackage rec {
 
   useDune2 = true;
 
-  propagatedBuildInputs = [ checkseum cmdliner git-unix yaml
-    irmin irmin-fs irmin-git irmin-graphql irmin-http irmin-mem irmin-pack irmin-watcher
+  propagatedBuildInputs = [
+    checkseum cmdliner git-unix yaml fpath
+    irmin irmin-fs irmin-git irmin-graphql irmin-http
+    irmin-pack irmin-watcher
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/irmin/unix.nix:18]:
```suggestion
    irmin-pack irmin-watcher git-cohttp-unix
```

File: pkgs/development/ocaml-modules/irmin/unix.nix [hunk 3]
```diff
@@ -1,7 +1,7 @@
 { lib, buildDunePackage
-, checkseum, cmdliner, git-unix, yaml
-, irmin, irmin-fs, irmin-git, irmin-graphql, irmin-http, irmin-mem, irmin-pack, irmin-watcher
-, irmin-test
+, checkseum, cmdliner, git-unix, yaml, fpath
+, irmin, irmin-fs, irmin-git, irmin-graphql, irmin-http
+, irmin-pack, irmin-watcher, irmin-test
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/irmin/unix.nix:4]:
```suggestion
, irmin-pack, irmin-watcher, irmin-test, git-cohttp-unix
```

**Comment by sternenseemann** [comment]
Incorporated your suggestions, good finds! I've also squashed all updates of the same packages as rebasing got complicated.

**Comment by sternenseemann** [comment]
Result of `nixpkgs-review pr 110184` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>ocamlPackages.wodan-irmin</li>
  </ul>
</details>
<details>
  <summary>28 packages built:</summary>
  <ul>
    <li>ocamlPackages.carton</li>
    <li>ocamlPackages.carton-git</li>
    <li>ocamlPackages.carton-lwt</li>
    <li>ocamlPackages.decompress</li>
    <li>ocamlPackages.duff</li>
    <li>ocamlPackages.encore</li>
    <li>ocamlPackages.git</li>
    <li>ocamlPackages.git-cohttp</li>
    <li>ocamlPackages.git-cohttp-mirage</li>
    <li>ocamlPackages.git-cohttp-unix</li>
    <li>ocamlPackages.git-unix</li>
    <li>ocamlPackages.imagelib</li>
    <li>ocamlPackages.index</li>
    <li>ocamlPackages.irmin</li>
    <li>ocamlPackages.irmin-chunk</li>
    <li>ocamlPackages.irmin-containers</li>
    <li>ocamlPackages.irmin-fs</li>
    <li>ocamlPackages.irmin-git</li>
    <li>ocamlPackages.irmin-graphql</li>
    <li>ocamlPackages.irmin-http</li>
    <li>ocamlPackages.irmin-layers</li>
    <li>ocamlPackages.irmin-mirage</li>
    <li>ocamlPackages.irmin-mirage-git</li>
    <li>ocamlPackages.irmin-mirage-graphql</li>
    <li>ocamlPackages.irmin-pack</li>
    <li>ocamlPackages.irmin-test</li>
    <li>ocamlPackages.irmin-unix</li>
    <li>ocamlPackages.ppx_irmin</li>
  </ul>
</details>

File: pkgs/development/ocaml-modules/irmin/pack.nix [hunk 4]
```diff
@@ -9,9 +11,9 @@ buildDunePackage rec {
 
   useDune2 = true;
 
-  propagatedBuildInputs = [ index irmin ocaml_lwt ];
+  propagatedBuildInputs = [ index irmin irmin-layers ocaml_lwt fpath ];
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/irmin/pack.nix:14]: `fpath` need not be propagated.

Comment (sternenseemann) [reply pkgs/development/ocaml-modules/irmin/pack.nix]: Done!

**Review by vbgl — APPROVED** [review]

**Pull request merged**